cabal.project

```diff
@@ -0,0 +1,28 @@
+packages:
+  csound-catalog
+  csound-controllers
+  csound-expression
+  csound-expression-dynamic
+  csound-expression-opcodes
+  csound-expression-typed
+  csound-sampler
+  gen-csound-opcodes
+
+-- https://github.com/haskell-crypto/cryptonite/pull/354
+source-repository-package
+  type: git
+  location: https://github.com/josephcsible/cryptonite
+  tag: 3b081e3ad027b0550fc87f171dffecbb20dedafe
+
+-- https://github.com/haskell-foundation/foundation/pull/555
+source-repository-package
+  type: git
+  location: https://github.com/TomMD/foundation
+  subdir: basement
+  tag: 0bb195e1fea06d144dafc5af9a0ff79af0a5f4a0
+
+-- https://github.com/vincenthz/hs-memory/pull/87
+source-repository-package
+  type: git
+  location: https://github.com/tfausak/hs-memory
+  tag: 3cf661a8a9a8ac028df77daa88e8d65c55a3347a
```

Comment:
Best not merge until these three packages (which notably all have the same maintainer...) are updated on Hackage.

Reply:
Thanks for update! I wonder why on earth I need cryptonite in such a library. Probably it's due to hashing functions, maybe it will be worth to investigate light weight alternatives

Reply:
gen-csound-opcodes -> http-conduit -> http-client-tls -> connection -> x509-system -> x509-store -> x509 -> cryptonite

the size of the dependency graph did explode? See these graphics, built with https://hackage.haskell.org/package/cabal-plan

![origin/master](https://www.imn.htwk-leipzig.de/~waldmann/etc/csound-expression/deps-origin-master.png)

![georgefst/master](https://www.imn.htwk-leipzig.de/~waldmann/etc/csound-expression/deps-georgefst-master.png)

Reply:
Aha, the dependency graph is much smaller without `gen-csound-opcodes`! And the package overrides aren't necessary at all for all the other packages.

It looks like `gen-csound-opcodes` is some kind of code-gen tool that gets run manually? At least, it isn't a transitive dependency of the `csound-expression` library. So actually, it would be possible to make a Hackage release with GHC 9.2 support without waiting for `cryptonite` etc.

Anyway, it's possible there is a more lightweight `http` stack, but it's less of an issue than we thought, given that `csound-expression` doesn't actually depend on it. And being held up on updating to 9.2 by these particular libs is certainly not unique: https://github.com/haskell-foundation/foundation/pull/555#issuecomment-983107899.

Reply:
Csound.Gen.Parser.downloadDesc makes a http request, is called from app/Prepare.hs, see README.md. I did not know about this. Amazing! (not the download itself, but the parsing and code generation.)

Reply:
Thanks @jwaldmann ! Yeah it's auto generated from online Csound docs. The whole package `csound-expression-opcodes` is generated with gen-opcodes. Csound has 1000+ opcodes and you can imagine what a pain is to update it manually if some core feature changes. So I did a web-crawler and code generator for that.

Ofcourse gen-opcodes has more deps than csound-xxx packages and it's used manually only from the stack version. There is no need to account for it's deps.

Reply:
I'm curious as to how the http stack is actually used.

> Anyway, it's possible there is a more lightweight http stack, but it's less of an issue than we thought, given that csound-expression doesn't actually depend on it.

It sounds like it might not. I only stumbled on this project today so I may be not adequately oriented to it, but I'm somewhat surprised that http is anywhere in an api that concerns itself with signal generation/processing.

Reply:
@ProofOfKeags It's  used to download online Csound docs and generate wrapper haskell code for it's 1000+ functions out of docs. It's not used in the package itself, only used to generate `csound-expression-opcodes` package

Reply:
> Best not merge until these three packages (which notably all have the same maintainer...) are updated on Hackage.

The time has come. [f96c5a1](https://github.com/spell-music/csound-expression/pull/68/commits/f96c5a11991ecd5ac3925363c895b3250943a2dc)